File: Ada/livro-ds/floresta.c
/* (C) 1999, Arthur Vargas Lopes */
/* Arquivo fonte Floresta.c      */

#include "Floresta.h"

/* Algoritmo 10.15 */
Floresta Inicializa(char Fn[])
{
	int      I;
	Floresta F;

	if ((F = (Floresta) calloc(1,sizeof(FlorestaSIZE))) == NULL)  
		return NULL;

	/* Cria árvore */
	if ((F->Tfp = fopen(Fn,"wb+")) == NULL)  
		return NULL;

	F->Em_uso    = 0;
	F->Deletados = 0;
	F->Compacte  = 0;

	strcpy(F->Tfn,Fn);

	for(I = 0; I < FLORESTA; ++I) 
		F->Raiz[I] = -2L;

	if (fwrite(F,sizeof(FlorestaSIZE),1,F->Tfp) != 1)  
		return NULL;
	
	if (fclose(F->Tfp) != 0)
		return NULL;

	if ((F->Tfp = fopen(Fn,"rb+")) == NULL)  
		return NULL;

	return F;

} /* Inicializa */

/* Algoritmo 10.16 */
Floresta Abre(char Fn[])
{
	double   Sit;
	Floresta F;
	FILE     *Ftemp;

	if ((F = (Floresta) calloc(1,sizeof(FlorestaSIZE))) == NULL)  
		return NULL;

	if ((Ftemp = fopen(Fn,"rb+")) == NULL)  
		return NULL;

	rewind(Ftemp);

	if (fread(F,sizeof(FlorestaSIZE),1,Ftemp) != 1)  
		return NULL;

	F->Tfp = Ftemp;

	if (F->Em_uso != 0)
		Sit = F->Deletados * 100.0 / F->Em_uso;

	if (Sit > 30.0 || F->Deletados > F->Em_uso)  
		F->Compacte = 1;
	else
		F->Compacte = 0;

	return F;

} /* Abre */

/* Algoritmo 10.17 */
int Fecha(Floresta F)
{
	rewind(F->Tfp);

	if (fwrite(F,sizeof(FlorestaSIZE),1,F->Tfp) != 1)  
		return 1;

	fflush(F->Tfp);

	if (fclose(F->Tfp) != 0)
		return 2;

	free(F);

	return 0;

} /* Fecha */

/* Algoritmo 10.18 */
int Hash(char *S)
{
	int          I;
	unsigned int Val = 0;

/* utilize este comando para tornar a floresta ordenada */
	/*	return *S; */

	for(I = 1; I <= strlen(S); ++I)
		Val = Val + (*S++ * I);

	return Val % FLORESTA;

} /* Hash */

/* Algoritmo 10.19 */
long Seek(Floresta F, long O, int M)
{
	long K;

	switch(M)  {
		case 0 : 
			if (fseek(F->Tfp, O, M) != 0) 
				break;
			if (ftell(F->Tfp) != O)
				break;
			return O;
		case 1 :
		case 2 :
			K = fseek(F->Tfp, O, M);
			if (K != 0) 
				break;
			return ftell(F->Tfp);
		default:
			return -1L;
		}

	return -2L;

} /* Seek */

/* Algoritmo 10.20 */
int Le(Floresta F, TKPTR Nodo, long W)
{
	if (Seek(F,W,0) < 0L)
		return 1;

	if (fread(Nodo,TREESIZE,1,F->Tfp) != 1)   
		return 2;

	if (fread(Nodo->Key, Nodo->Len,1,F->Tfp) != 1)  
		return 3;

	Nodo->Key[Nodo->Len] = '\0';

	return 0;

} /* Le */

/* Algoritmo 10.21 */
int Grava(Floresta F, TKPTR Nodo, long W)
{
	if (Seek(F,W,0) < 0L)
		return 1;

	Nodo->Len = strlen(Nodo->Key);

	if (fwrite(Nodo,TREESIZE,1,F->Tfp) != 1)   
		return 2;

	if (fwrite(Nodo->Key,(int) Nodo->Len,1,F->Tfp) != 1)  
		return 3;

	return 0;

} /* Grava */

/* Algoritmo 10.22 */
long Prepara(Floresta F, char *K, long Tag)
{
	TKSIZE Nodo;
	long   X;

	if ((X = Seek(F,0L,2)) < 0L)
		return -1L;

	Nodo.LLink  = -2L;
	Nodo.RLink  = -2L;
	Nodo.Tag    = Tag;
	Nodo.Status = 'I';
	Nodo.Len    = strlen(K);

	F->Em_uso += 1;
	strcpy(Nodo.Key,K);

	if (fwrite(&Nodo,TREESIZE,1, F->Tfp) != 1)  
		return -2L;

	if (fwrite(Nodo.Key, (int) Nodo.Len,1,F->Tfp) != 1)  
		return -3L;

	return X;

} /* Prepara */

/* Algoritmo 10.23 */
long Insere(Floresta F, char *K, long Tag)
{
	TKSIZE Nodo;
	long   W , X;
	int    I;

	I = Hash(K);

	if ((W = F->Raiz[I]) < 0L)  {

		F->Raiz[I] = Prepara(F,K,Tag);
		if (F->Raiz[I] < 0L) 
			return -4L;

		if (Seek(F,0L,0) < 0L)
			return -3L;

		if (fwrite(F,sizeof(FlorestaSIZE),1,F->Tfp) != 1)  
			return -1L;

		fflush(F->Tfp);

		return(F->Raiz[I]);
		}

	while ( 1 )  {

		if (Le(F,&Nodo,W) != 0)
			return -5L;

		switch(strcmp(K,Nodo.Key))  {
			case -1 : 
				if (Nodo.LLink < 0L)  {
					X = Nodo.LLink = Prepara(F,K,Tag);
					if (X < 0L)
						return -4L;
					if (Grava(F,&Nodo,W) != 0)
						return -6L;
					return X;
					}
				W = Nodo.LLink;
				break;

			case  0 :
				if (Nodo.Status != 'D')  
					return -2L;
				F->Deletados     -= 1;
				Nodo.Status          = 'I';
				Nodo.Tag             = Tag;
				if (Grava(F,&Nodo,W) != 0)
					return -6L;
				return W;

			default :
				if (Nodo.RLink < 0L)  {
					X = Nodo.RLink = Prepara(F,K,Tag);
					if (X < 0L)
						return -4L;
					if (Grava(F,&Nodo,W) != 0)
						return -6L;
					return X;
					}
				W = Nodo.RLink;
				break;
			}
		}
} /* Insere */

/* Algoritmo 10.24 */
long Consulta(Floresta F, char *K)
{
	TKSIZE Nodo;
	long   X;

	X = F->Raiz[Hash(K)];

	while ( 1 )  {
		if ( X < 0L) 
			return -1L;

		if (Le(F,&Nodo,X))
			return -2L;

		switch(strcmp(K,Nodo.Key))  {
			case -1 : 
				X = Nodo.LLink; 
				break;
			case  0 : 
				if (Nodo.Status == 'D') 
					return -1L;
				return(Nodo.Tag);
			default : 
				X = Nodo.RLink; 
				break;
			}
		}
} /* Consulta */

/* Algoritmo 10.25 */
int Remove(Floresta F, char *K)
{
	TKSIZE Nodo;
	long   X;

	X = F->Raiz[Hash(K)];

	while ( 1 )  {

		if (X < 0L) 
			return 1;

		if (Le(F,&Nodo,X)) 
			return 2;

		switch(strcmp(K,Nodo.Key))  {
			case -1 : 
				X = Nodo.LLink; 
				break;

			case  0 : 
				if (Nodo.Status == 'D') 
					return 3;

				Nodo.Status      = 'D';
				F->Deletados += 1;
				if (Grava(F,&Nodo,X) != 0)
					return 4;
				return 0;

			default : 
				X = Nodo.RLink; 
				break;
			}
		}
} /* Remove */

/* Algoritmo 10.26 */
int Tempo_De_Compactar(Floresta F)
{
	return F->Compacte;

} /* Tempo_De_Compactar */

/* Algoritmo 10.27 */
int Squash(char Fn[])
{
	char     Tn[101];
	long     K, Fim;
	TKSIZE   Nodo;
	Floresta F;
	Floresta Temp;

	if ((F = Abre(Fn)) == NULL)
		return 1;

	if ((Temp = Inicializa("TREEBAK.BAK")) == NULL)
		return 1;

	K = sizeof(FlorestaSIZE);
	if (fseek(F->Tfp, 0L, 2) != 0)
		return 2;

	Fim = ftell(F->Tfp);

	while ( K < Fim )   {
		if (Le(F,&Nodo,K) != 0)
			return 2;
		if (Nodo.Status != 'D')
			if (Insere(Temp,Nodo.Key,Nodo.Tag) < 0L)
				return 3;
		K = ftell(F->Tfp);
		}

	strcpy(Tn,F->Tfn);
	strcpy(Temp->Tfn,Tn);
	
	if (Fecha(Temp))
		return 4;

	if (Fecha(F))
		return 4;

	/* Remove árvore antiga */
	unlink(Tn);

	if (rename("TREEBAK.BAK",Tn) != 0)  
		return 5;

	return 0;

} /* Squash */

/* Algoritmo 10.28 */
void Imprime(Floresta F)
{ 
	int I;

	for(I = 0; I < FLORESTA; ++I)
		Central(F,F->Raiz[I]);

} /* Imprime */

/* Algoritmo 7.2 */
void Central(Floresta F, long K)
{ 
	TKSIZE Nodo;
	int    E;

	if (K <= 0L)
		return;

	if ((E = Le(F,&Nodo,K)) != 0) {
		printf("*** Erro em Central(%ld) E = %d\n", K, E);
		exit(1);
		}

	Central(F,Nodo.LLink);

	if (Nodo.Status != 'D')
		printf("%-40s %ld\n", Nodo.Key, Nodo.Tag);

	Central(F,Nodo.RLink);

} /* Central */

File: Ada/livro-ds/floresta.pas
{ (C) 1999, Arthur Vargas Lopes }
{ Arquivo fonte Floresta.Pas    }
PROGRAM Floresta;

	USES Crt;

CONST

	Floresta = 256;

	WLen	 = 30;

	WFn      = 128;

TYPE

	TKSize = RECORD
		LLink  : INTEGER;
		RLink  : INTEGER;
		Tag    : INTEGER;
		Status : CHAR;    { 'I' em uso; 'D' = deletado }
		Key    : STRING[WLen];
		END;

	FlorestaSize = RECORD
		Tfp       : FILE OF TKSize;
		Raiz      : ARRAY[1..Floresta] OF LONGINT;
		Em_uso    : INTEGER;
		Deletados : INTEGER;
		Nodos     : INTEGER;
		Compacte  : BOOLEAN;
		Tfn       : STRING[WFn];
		END;

	Florest = RECORD
		F   : FlorestaSize;
		Thp : FILE OF FlorestaSize;
		Thn : STRING[WFn];
		END;
VAR

	Header : Florest;
	K      : INTEGER;

	PROCEDURE Faz_Header(Fn : STRING; VAR H : Florest);
	VAR
		I : INTEGER;
	BEGIN
		H.Thn := Fn;
		I     := 1;

		WHILE Fn[I] <> '.' DO
			I := I + 1;

		H.Thn[I+1] := 'H';
		H.Thn[I+2] := 'D';
		H.Thn[I+3] := 'R';

	END; { Faz_Header }

	{ Algoritmo 10.15 }
	PROCEDURE Inicializa(Fn : STRING;VAR H : Florest);
	VAR
		I  : INTEGER;
	BEGIN

		{ Cria árvore }
		Faz_Header(Fn,H);
		Assign(H.Thp,H.Thn);
		ReWrite(H.Thp);

		H.F.Em_uso    := 0;
		H.F.Deletados := 0;
		H.F.Compacte  := FALSE;

		H.F.Tfn       := Fn;

		FOR I := 1 TO Floresta DO
			H.F.Raiz[I] := -2;

		Write(H.Thp,H.F);

		Close(H.Thp);

		Reset(H.Thp);

		Assign(H.F.Tfp,Fn);
		ReWrite(H.F.Tfp);

	END; { Inicializa }

	{ Algoritmo 10.16 }
	PROCEDURE Abre(Fn : STRING; VAR H : Florest);
	VAR
		Sit : REAL;
	BEGIN
		Faz_Header(Fn,H);
		Assign(H.Thp,H.Thn);
		Reset(H.Thp);

		Seek(H.Thp,1);
		Read(H.Thp,H.F);

		Assign(H.F.Tfp,Fn);
		Reset(H.F.Tfp);

		IF (H.F.Em_uso <> 0) THEN
			Sit := H.F.Deletados * 100.0 / H.F.Em_uso;

		IF (Sit > 30.0) OR (H.F.Deletados > H.F.Em_uso) THEN
			H.F.Compacte := TRUE
		ELSE
			H.F.Compacte := FALSE;

	END; { Abre }

	{ Algoritmo 10.17 }
	PROCEDURE Fecha(VAR H : Florest);
	BEGIN
		Seek(H.Thp,1);

		Close(H.F.Tfp);

		Write(H.Thp,H.F);

		Close(H.Thp);

	END; { Fecha }

	{ Algoritmo 10.18 }
	FUNCTION Hash(S : STRING) : INTEGER;
	VAR
		I   : INTEGER;
		VAL : INTEGER;
	BEGIN
		Val := 0;

		{ utilize este comando para tornar a floresta ordenada }
		{ Hash := S[1]; }

		FOR I := 1 TO Length(S) DO
			Val := Val + (Ord(S[I]) * I);

		Hash := Abs(Val MOD Floresta) + 1;

	END; { Hash }

	{ Algoritmo 10.20 }
	PROCEDURE Le(VAR H    : Florest;
		     VAR Nodo : TKSize;
		     W        : INTEGER);
	BEGIN
		Seek(H.F.Tfp,W);

		Read(H.F.Tfp,Nodo);

	END; { Le }

	{ Algoritmo 10.21 }
	PROCEDURE Grava(VAR H : Florest; Nodo : TKSize; W : INTEGER);
	BEGIN
		Seek(H.F.Tfp,W);

		Write(H.F.Tfp,Nodo);

	END; { Grava }

	{ Algoritmo 10.22 }
	FUNCTION Prepara(VAR H : Florest;
			 K     : STRING;
			 Tag   : INTEGER) : INTEGER;
	VAR
		Nodo : TKSize;
		X    : INTEGER;
	BEGIN

		X := H.F.Em_uso + 1;

		Seek(H.F.Tfp,X);

		Nodo.LLink  := -2;
		Nodo.RLink  := -2;
		Nodo.Tag    := Tag;
		Nodo.Status := 'I';

		H.F.Em_uso  := H.F.Em_uso + 1;
		Nodo.Key    := K;

		Write(H.F.Tfp,Nodo);

		Prepara := X;

	END; { Prepara }

	{ Algoritmo 10.23 }
	PROCEDURE Insere(VAR H : Florest; K : STRING; Tag : INTEGER);
	VAR
		Nodo : TKSize;
		I, W : INTEGER;
	BEGIN

		I := Hash(K);

		W := H.F.Raiz[I];

		IF W < 0 THEN BEGIN

			H.F.Raiz[I] := Prepara(H,K,Tag);

			Seek(H.Thp,1);

			Write(H.Thp,H.F);

			END
		ELSE BEGIN

			WHILE W >= 0 DO BEGIN

				Le(H,Nodo,W);

				IF K < Nodo.Key THEN
					IF Nodo.LLink < 0 THEN BEGIN
						Nodo.LLink := Prepara(H,K,Tag);
						Grava(H,Nodo,W);
						W := -1;
						END
					ELSE
						W := Nodo.LLink
				ELSE IF K = Nodo.Key THEN
					IF Nodo.Status <> 'D' THEN
						W := -1
					ELSE BEGIN
						H.F.Deletados := H.F.Deletados - 1;
						Nodo.Status   := 'I';
						Nodo.Tag      := Tag;
						Grava(H,Nodo,W);
						END
				ELSE
					IF Nodo.RLink < 0 THEN BEGIN
						Nodo.RLink := Prepara(H,K,Tag);
						Grava(H,Nodo,W);
						END
					ELSE
						W := Nodo.RLink;
				END;
			END;
	END; { Insere }

	{ Algoritmo 10.24 }
	FUNCTION Consulta(H : Florest; K : STRING) : INTEGER;
	VAR
		Nodo : TKSize;
		X    : INTEGER;
	BEGIN
		Consulta := 0;

		X := H.F.Raiz[Hash(K)];

		WHILE  X >= 1 DO BEGIN

			Le(H,Nodo,X);

			IF K < Nodo.Key THEN
				X := Nodo.LLink
			ELSE IF K = Nodo.Key THEN
				IF Nodo.Status <> 'D' THEN BEGIN
					Consulta := Nodo.Tag;
					X        := 0;
					END
				ELSE
					X := 0
			ELSE
				X := Nodo.RLink;
			END;
	END; { Consulta }

	{ Algoritmo 10.25 }
	FUNCTION Remove(VAR H : Florest; K : STRING) : BOOLEAN;
	VAR
		Nodo : TKSize;
		X    : INTEGER;
	BEGIN
		Remove := FALSE;

		X      := H.F.Raiz[Hash(K)];

		WHILE X >= 1 DO BEGIN

			Le(H,Nodo,X);

			IF K < Nodo.Key THEN
				X := Nodo.LLink
			ELSE IF K = Nodo.Key THEN
				IF Nodo.Status = 'D' THEN
					X             := 0
				ELSE BEGIN
					Nodo.Status   := 'D';
					H.F.Deletados := H.F.Deletados + 1;
					Grava(H,Nodo,X);
					Remove        := TRUE;
					X             := 0;
					END
			ELSE
				X := Nodo.RLink;
			END;
	END; { Remove }

	{ Algoritmo 10.26 }
	FUNCTION Tempo_De_Compactar(H : Florest) : BOOLEAN;
	BEGIN
		Tempo_De_Compactar := H.F.Compacte;

	END; { Tempo_De_Compactar }


	{ Algoritmo 10.27 }
	PROCEDURE Squash(Fn : STRING);
	VAR
		Tn1, Tn2 : STRING[WFn];
		Nodo     : TKSize;
		H, Temp  : Florest;
		K        : INTEGER;

	BEGIN
		Abre(Fn,H);

		Inicializa('TREEBAK.BAK',Temp);

		K := 1;
		WHILE NOT Eof(H.F.Tfp) DO BEGIN
			Le(H,Nodo,K);
			IF Nodo.Status <> 'D' THEN
				Insere(Temp,Nodo.Key,Nodo.Tag);
			K := K + 1;
		END;

		Tn1        := H.F.Tfn;
		Temp.F.Tfn := Tn1;
		Tn2        := H.Thn;

		Fecha(Temp);

		Fecha(H);

		{ Remove árvore antiga }
		Erase(H.F.Tfp);
		Erase(H.Thp);

		Rename(Temp.F.Tfp,Tn1);
		Rename(Temp.Thp,Tn2);


	END; { Squash }

	{ Algoritmo 7.2 }
	PROCEDURE Central(H : Florest; K: INTEGER);
	VAR
		Nodo : TKSize;
		I    : INTEGER;
	BEGIN

		IF K > 0 THEN BEGIN
			Le(H,Nodo,K);

			Central(H,Nodo.LLink);

			IF Nodo.Status <> 'D' THEN BEGIN
				Write(Nodo.Key);
				FOR I := Length(Nodo.Key) + 1 TO 41 DO
					Write(' ');
				Writeln(Nodo.Tag);
				END;

			Central(H,Nodo.RLink);
			END;

	END; { Central }

	{ Algoritmo 10.28 }
	PROCEDURE Imprime(H : Florest);
	VAR
		I : INTEGER;
	BEGIN
		FOR I := 1 TO Floresta DO
			Central(H,H.F.Raiz[I]);

	END; { Imprime }
BEGIN
	Inicializa('ADISCO.DAT',Header);

	Insere(Header,'Fulano de Tal',1010);
	Insere(Header,'Fulana de Tal',1020);
	Insere(Header,'Ciclano de Tal',1030);
	Insere(Header,'Beltrano',1040);
	Insere(Header,'Pertrano',1050);

	Imprime(Header);

	Fecha(Header);

	Abre('ADISCO.DAT',Header);

	K := Consulta(Header,'Ciclano de Tal');
	IF K > 0 THEN
		Writeln('Ciclano de Tal tem tag ', K)
	ELSE
		Writeln('Nao encontrou Ciclano de Tal');

	IF Remove(Header,'Ciclano de Tal') THEN
		Writeln('Ciclano de Tal foi removido')
	ELSE
		Writeln('Nao removeu Ciclano de Tal');

	K := Consulta(Header,'Ciclano de Tal');
	IF K > 0 THEN
		Writeln('Ciclano de Tal tem tag ', K)
	ELSE
		Writeln('Nao encontrou Ciclano de Tal');

	Fecha(Header);

	Squash('ADISCO.DAT');

	Abre('ADISCO.DAT',Header);

	Imprime(Header);

	Fecha(Header);

END. { Floresta }
